ingestion/ingest-google-docs.py

```diff
@@ -0,0 +1,134 @@
+from git import Repo
+
+SCOPES = ['https://www.googleapis.com/auth/drive.metadata.readonly']
+SCOPES_DOC = ['https://www.googleapis.com/auth/documents.readonly']
+DISCOVERY_DOC = ('https://docs.googleapis.com/$discovery/rest?'
+                         'version=v1')
+
+ROOT_PATH = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..'))
+GIT_REPO_PATH = f'{ROOT_PATH}/.git'
+COMMIT_MESSAGE = 'Update shared repository'
+
+def git_push():
+    repo = Repo(GIT_REPO_PATH)
+    repo.index.add(["shared_directory"])
+    repo.index.commit(COMMIT_MESSAGE)
+    repo.git.push('origin', 'feature_docs_api') 
+
+def read_paragraph_element(element):
+    """Returns the text in the given ParagraphElement.
+
+        Args:
+            element: a ParagraphElement from a Google Doc.
+    """
+    text_run = element.get('textRun')
+    if not text_run:
+        return ''
+    return text_run.get('content')
+
+
+def read_strucutural_elements(elements):
+    """Recurses through a list of Structural Elements to read a document's text where text may be
+        in nested elements.
+
+        Args:
+            elements: a list of Structural Elements.
+    """
+    text = ''
+    for value in elements:
+        if 'paragraph' in value:
+            elements = value.get('paragraph').get('elements')
+            for elem in elements:
+                text += read_paragraph_element(elem)
+        elif 'table' in value:
+            # The text in table cells are in nested Structural Elements and tables may be
+            # nested.
+            table = value.get('table')
+            for row in table.get('tableRows'):
+                cells = row.get('tableCells')
+                for cell in cells:
+                    text += read_strucutural_elements(cell.get('content'))
+        elif 'tableOfContents' in value:
+            # The text in the TOC is also in a Structural Element.
+            toc = value.get('tableOfContents')
+            text += read_strucutural_elements(toc.get('content'))
+    return text
+
+
+def run():
+    # Find ids of all Google docs in the raw Serge folder
+    store = file.Storage('secrets/token_read.json')
+    page_token = None   
+    creds = None
+    flow = client.flow_from_clientsecrets('secrets/credentials.json', SCOPES)
+    if os.path.exists('secrets/token_read.pickle'):
+        with open('secrets/token_read.pickle', 'rb') as token:
+            creds = pickle.load(token)
+        # If there are no (valid) credentials available, let the user log in.
+    if not creds or creds.invalid:
+        if creds and creds.expired and creds.refresh_token:
+            creds.refresh(Request())
+        else:
+            flow = client.flow_from_clientsecrets('secrets/credentials.json', SCOPES)
+            creds = tools.run_flow(flow, store)
+        # Save the credentials for the next run
+        with open('secrets/token_read.pickle', 'wb') as token:
+            pickle.dump(creds, token)

```

Comment:
Why do we pickle these credentials for the next run, but not creds_doc below?

Btw, can we move this credential logic to a helper method?

Reply:
I refactored this all to use the same authentication flow for both docs and drive and made it a method outside of `run()`